torch.sort: Optimize memory usage with (dtype_indices: ScalarType, dynamic_indices_dtype: bool) options

voidbag opened the pull request:
Fixes #147628

The static dtype of indices was kLong(64bit), consuming exccessive memory.
It reduces memory usage, determining the dtype of indices dynamically.
The dtype is one of Byte, UInt16, UInt32 and UInt64.

- ~This PR makes at::arange support uint16, uint32 and uint64 (`at::arange( uint(16|32|64) )->uint(16|32|64)`~
- ~It also changes existing behavior of at::arange, `at::arange(uint8)->int64` into  `at::arange(uint8)->uint8`~
- This PR makes at::linspace support uint16, uint32 and uint64



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @yf225

voidbag committed:
Sort: optimize CPU/GPU memory usage with dynamic dtype of indices

The static dtype of indices was kLong(64bit), consuming exccessive memory.

This feature reduces memory usage, determining the dtype of indices dynamically.

The dtype is one of at::ScalarType(s) Byte, UInt16, UInt32 and UInt64.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on aten/src/ATen/native/cpu/SortingKernel.cpp:
```diff
@@ -179,47 +194,52 @@ static void sort_kernel(
-    parallel_sort1d_kernel(values, indices);
-    return;
-  }
+    if (can_use_radix_sort(values, descending)) {
+      using signed_t = UnsignedType<scalar_t>::signed_type; //FBGEMM explicit template instantiations don't contain unsigned type.
+      auto* const indices_ptr = reinterpret_cast<signed_t *>(indices.data_ptr<scalar_t>());
+      parallel_sort1d_kernel<signed_t>(values, indices_ptr);
```
Explicit template instantiation in fbgemm::radix_sort_parallel only supports uint8_t among unsigned dtype(s). That's why unsigned values should be reinterpret_casted to signed values.

ngimel commented:
As mentioned in the issue, making output dtype dependent on the size of the input tensor is confusing and bc-breaking, so we won't go with that. However, doing intermediate computations with a narrower dtype and copying to int64 will still result in memory (and possible perf) improvement, so if you are interested you can update your PR to do those non-bc-breaking changes.

voidbag committed:
linspace: support UInt16, UInt32, UInt64

It's for dynamic dtype of at::sort.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
sort: add 'dynamic_indices_dtype'

The dtype of indices is selected among uint8, 16, 32 and 64 automatically, if the option is set.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
The option `dynamic_indices_type` is added to at::sort.
If the option is set, the function returns indices (uint8, uint16, uint32 or uint64)

voidbag commented:
After I viewed the CI log, I'll apply the option to at::argsort

ngimel commented:
This significantly increases number of template instantiations for sorting, and it is already one of the worst offenders for compile time. 
cc @malfet, @albanD

voidbag committed:
sort: handle onnx test failures

Handle Onnx Test Failure
- Handles onnx test failure by adding parameters in
  torch/onnx/symbolic_opset(9|10|11).py

Handle Backwards Compatibility Issue
- Tensor.sort(true, 0, true) in torchvision calls
  Tensor.sort(int dim, bool descending, bool dynamic_indices_type)
  rather than Tensor.sort(optional<bool> stable, int dim, bool descending)
- It's resolved by changing the function's prototype
  as sort(int dim, bool descending, ScalarType indices_dtype,
          bool dynamic_indices_type)

Add `str == "at::kLong"` in python_arg_parser.cpp.
- `at::kLong` cannot be parsed in
  `python_(variable_methods|torch_functionsEverything|torch_functions_2)`.cpp
   in `autograd/generated` directory
- It's resolved by adding "at::kLong" or-condition to "torch.int64"

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented:
I changed the function prototype as (ScalarType indices_dtype, bool dynamic_indices_dtype),which resolves ambiguity issue in function overloads. And this patch resolves onnx test failures.

@ngimel 
Would you mind triggering CI?

voidbag commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -1260,7 +1260,7 @@ void FunctionParameter::set_default_str(const std::string& str) {
     if (str == "None") {
       default_scalartype = at::ScalarType::Undefined;
-    } else if (str == "torch.int64") {
+    } else if (str == "torch.int64" || str == "at::kLong") {
```
It's a little tricky.

But, It can handle below cases.
```
// torch/csrc/autograd/generated/python_variable_methods.cpp
// torch/csrc/autograd/generated/python_torch_functionsEverything.cpp
// torch/csrc/autograd/generated/python_torch_functions_2.cpp
static PythonArgParser parser({
   "sort(*, bool? stable, Dimname dim, bool descending=False)",
   "sort(*, bool? stable, int64_t dim=-1, bool descending=False, ScalarType indices_dtype=at::kLong, bool dynamic_indices_dtype=False)",
   "sort(Dimname dim, bool descending=False)",
   "sort(int64_t dim=-1, bool descending=False, ScalarType indices_dtype=at::kLong, bool dynamic_indices_dtype=False)",
}, /*traceable=*/true);
```

albanD commented:
Hey @voidbag , I wanted to chime in and take a step back on this.
While I think we all agree that this is a nice improvement and cool change, there are very large implications and operational issues from the native_functions.yaml change you're doing here. And I do expect it is going to take us several weeks (most likely month) to get a PR with the current scope ready for merge.
Concretely (and NOT asking you to make these changes), we would need to do TS upgrade properly, update the abi-stable shim properly, discuss with the broader maintainers the new arguments for such a commonly used function, update testing, update docs, etc. I don't expect it is worth spending so much time developing and reviewing such a change for the benefit we're going to get.

I would suggest the best way forward here to avoid these operational changes is either:
- If you need this improvement for your own work fast, I would suggest wrap that updated kernel into a custom op lib you can use as a standalone. Since you don't need to land this in core, you will be able to get that out as fast as you need.
- If you want to work on this PR, I would agree with @ngimel proposal above and avoid changing the signature in any way. Doing kernel-only change would give you most of the benefit without having to deal with the many operational issues of changing this signature.

Let me know what you think!

voidbag commented:
@albanD 
The signiture change I proposed requires much costs...
Then, I'll change the internal implementation of torch.sort without changing the signiture as you said on this PR.

I have a question. 
What if I open a new PR adding only `indices_dtype'`parameter(excluding dynamic_indices_dtype)? as soon as this PR is applied 
-  I might rewrite the torch.sort signiture and test/docs codes consistent with cumsum signature **dtype** convention `torch.cumsum(input, dim, *, dtype=None, out=None) → Tensor`. 
- The test code and docs of torch.sort with **indices_dtype** could follow the convention of cumsum signiture with **dtype**, which can reduce some engineering costs.

voidbag committed:
torch.sort: indices supports all integral dtypes

change dispatch macro to support all integral types of indices

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
Revert to origin signature of torch.sort

Revert to origin signature of torch.sort...

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag committed:
torch.sort: downcast indices and upcast to kLong

The process of downcasting and upcasting indices saves memory usage.
It doesn't require the signature of torch.sort.

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

ngimel commented:
The docker image errors will be solved with rebase, but win build error seems real

voidbag commented:
> The docker image errors will be solved with rebase, but win build error seems real

I think the win build failure is caused by `#ifdef in AT_DISPATCH_V2`.
```
  AT_DISPATCH_V2(indices.scalar_type(), "sort_kernel", AT_WRAP([&] {
#ifdef USE_FBGEMM
    if (can_use_radix_sort(values, descending)) {
      using signed_t = UnsignedType<scalar_t>::signed_type; //FBGEMM explicit template instantiations don't contain unsigned type.
      auto* const indices_ptr = reinterpret_cast<signed_t *>(indices.data_ptr<scalar_t>());
      parallel_sort1d_kernel<signed_t>(values, indices_ptr);
```

I'll try to resolve it.

voidbag committed:
SortingKernel.cpp: fix win build failure

remove `#ifdef in AT_DISPATCH_V2` for win build

Signed-off-by: Taeksang Kim <voidbag@gmail.com>

voidbag commented on aten/src/ATen/native/cpu/SortingKernel.cpp:
```diff
@@ -89,8 +89,26 @@ struct KeyValueCompDesc {
+TO_TYPE(int64_t, int64_t)
+
 static bool can_use_radix_sort(const TensorBase& values, const bool descending) {
+#ifdef USE_FBGEMM
```
I moved #ifdef in static function definition for win build

voidbag commented:
- Rebase has been completed.
- Suspicious code snippet about win build failure has been modified.

@ngimel
Would you mind triggering CI?

ngimel commented:
Apparently SortPairs compile time is already a problem #148936, so 4x-ing it is not the best approach. Can we limit instantiations to int32 only, or have some jiterator-like runtime solution?

voidbag commented:
I think we need not only int32 but also uint8 and uint16. The memory saved will be much larger if the internal dtype becomes smaller like uint8, uint16, as I already reported the effect in the above table.
Additionally, I hope supporting all integer types including `unsigned` makes it easier to add indices_dtype parameter in torch.sort signature in the future PR.

I don't know 'jiterator-like runtime solution' that you said, exactly. Can I have some reference?

And I'm also worried about how to change this PR, after #148936 ...

voidbag committed:
temporary commit to prevent bot from closing it

Signed-off-by: Taeksang Kim <voidbag@gmail.com>
